src/bundle.c

```diff
@@ -426,126 +429,157 @@ int remove_bundle(const char *bundle_name)
+			bad++;
+			goto out_free_mom;
+		}
 
-	/* deduplication needs file list sorted by filename, do so */
-	bundle_manifest->files = list_sort(bundle_manifest->files, file_sort_filename);
-	deduplicate_files_from_manifest(&bundle_manifest, current_mom);
+		/* check if bundle is required by another installed bundle */
+		struct list *reqd_by = NULL;
+		required_by(&reqd_by, *bundles, current_mom, 0);
+		if (reqd_by != NULL) {
+			struct list *iter;
+			char *bundle;
+			iter = list_head(reqd_by);
+			fprintf(stderr, "Error: bundle requested to be removed is required by the following bundles:\n");
+			printf("format:\n");
+			printf(" # * is-required-by\n");
+			printf(" #   |-- is-required-by\n");
+			printf(" # * is-also-required-by\n # ...\n\n");
+			while (iter) {
+				bundle = iter->data;
+				iter = iter->next;
+				fprintf(stderr, "%s", bundle);
+			}
 
-	fprintf(stderr, "Deleting bundle files...\n");
-	remove_files_in_manifest_from_fs(bundle_manifest);
+			list_free_list(reqd_by);
+			ret = EBUNDLE_REMOVE;
+			bad++;
+			goto out_free_mom;
+		}
 
-	fprintf(stderr, "Untracking bundle from system...\n");
-	rm_bundle_file(bundle_name);
+		current_mom->files = files_from_bundles(current_mom->submanifests);
+		current_mom->files = consolidate_files(current_mom->files);
 
-	fprintf(stderr, "Success: Bundle removed\n");
+		/* Now that we have the consolidated list of all files, load bundle to be removed submanifest */
+		ret = load_bundle_manifest(*bundles, subs, current_version, &bundle_manifest);
+		if (ret != 0 || !bundle_manifest) {
+			fprintf(stderr, "Error: Cannot load %s sub-manifest (ret = %d)\n", *bundles, ret);
+			bad++;
+			goto out_free_mom;
+		}
 
-	free_manifest(bundle_manifest);
-out_free_mom:
-	free_manifest(current_mom);
-out_free_curl:
-	telemetry(ret ? TELEMETRY_CRIT : TELEMETRY_INFO,
-		  "bundleremove",
-		  "bundle=%s\n"
-		  "current_version=%d\n"
-		  "result=%d\n",
-		  bundle_name,
-		  current_version,
-		  ret);
-	if (ret) {
-		fprintf(stderr, "Error: Bundle remove failed\n");
+		/* deduplication needs file list sorted by filename, do so */
+		bundle_manifest->files = list_sort(bundle_manifest->files, file_sort_filename);
+		deduplicate_files_from_manifest(&bundle_manifest, current_mom);
+
+		fprintf(stderr, "Deleting bundle files...\n");
+		remove_files_in_manifest_from_fs(bundle_manifest);
+
+		fprintf(stderr, "Untracking bundle from system...\n");
+		rm_bundle_file(*bundles);
+
+		fprintf(stderr, "Success: Bundle removed\n");
+
+		free_manifest(bundle_manifest);
+	out_free_mom:
+		free_manifest(current_mom);
+	out_free_curl:
+		telemetry(ret ? TELEMETRY_CRIT : TELEMETRY_INFO,
+			  "bundleremove",
+			  "bundle=%s\n"
+			  "current_version=%d\n"
+			  "result=%d\n",
+			  *bundles,
+			  current_version,
+			  ret);
+		if (ret) {
+			fprintf(stderr, "Error: Bundle remove failed\n");
+			retCode = ret;
+		}
+	}
+
+	if (bad > 0) {
+		fprintf(stderr, "%i bundle(s) of %i failed to remove\n", bad, total);
```

Comment:
I know you are not adding this line, but it would be nice to correct this message (in another commit) to something like: "%i bundle(s) of %i failed to be removed".
And to be honest I don't even know if my suggestion is correct.